### PR TITLE
Add a tech note for interfaces in Chapel

### PR DESCRIPTION
**Pull request by DanilaFe**
This locks down the behavior of interfaces. I've elected to only write about stable language features here, which was OK'ed with the team.

Reviewed by @ShreyasKhandekar -- thanks!

### CODE REVIEW

File: doc/rst/technotes/interfaces.rst [hunk 1]
```diff
@@ -0,0 +1,235 @@
+.. _readme-interfaces:
+
+.. default-domain:: chpl
+
+
+========================================
+Interface in the Chapel Standard Library
```
Comment (ShreyasKhandekar) [inline doc/rst/technotes/interfaces.rst:7]: Interfaces?

File: doc/rst/technotes/interfaces.rst [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+  a combination of the ``writeSerializable``, ``readDeserializable``, and
+  ``initDeserializable`` interfaces. For a type to implement ``serializable``,
+  it must satisfy the other three interfaces as well.
+
+A function or an associated type that serves to satisfy a requirement of an
+interface is typically called a `witness`.
+
+When a type is marked as implementing an interface, the Chapel compiler will
+ensure that it satisfies all of the interface's requirements. The compiler will
+do so by checking the current scope, and resolving any functions that fit the
```
Comment (ShreyasKhandekar) [inline doc/rst/technotes/interfaces.rst:90]: drop comma

File: doc/rst/technotes/interfaces.rst [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+The Chapel compiler attempts to automatically generate certain methods for types,
+to preclude the need for users to implement their own. For instance, records
+without user-defined ``==`` and ``!=`` operators that lack a ``hash`` method
+are automatically supplied with a compiler-generated ``hash``. This serves to
+reduce the boilerplate for types that don't require custom logic. For some
+interfaces (described in this section), the compiler will also automatically
+generate an implementation if the required methods were generated.
+
+For or instance, consider the following snippet:
```
Comment (ShreyasKhandekar) [inline doc/rst/technotes/interfaces.rst:186]: ~or~

File: doc/rst/technotes/interfaces.rst [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+   var D: domain(R);
+   writeln(D);
+   writeln(new R());
+
+In the snippet, a newly-defined record ``R`` could be used in an associative
+domain and printed out using ``writeln``. Using these features requires the
+record to be hashable and serializable, respectively. However, the user was
+not required to manually implement ``hashable`` and ``writeSerializable``.
+
+Generally (and with the exception of transitory measures; see `Migration Support`_),
```
Comment (ShreyasKhandekar) [inline doc/rst/technotes/interfaces.rst:203]: and with the exception of -> except for

File: doc/rst/technotes/manage.rst [hunk 5]
```diff
@@ -19,20 +19,22 @@ keyword to open the statement instead of ``with``:
 The ``manage`` statement accepts a `manager` (the `myManager()` call
 in the above example). The statement calls a special method on the
 manager which lets it perform actions before executing the managed
-block.
+block. In order of a type to be recognized as a valid manager, it must
```
Comment (ShreyasKhandekar) [inline doc/rst/technotes/manage.rst:22]: In order of a type -> For a type...

File: doc/rst/technotes/interfaces.rst [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+========================================
+Interface in the Chapel Standard Library
+========================================
+
+Interfaces in Chapel are similar to a feature of the same name in object-oriented
+languages such as Java, and to `traits in Rust <https://doc.rust-lang.org/book/ch10-02-traits.html>`_.
+They provide a way to mark a type as exhibiting certain behavior or implementing
+certain functionality. For instance, a ``hashable`` interface marks a type as
+specifying a ``hash`` method (which would allow it to be used in
+:ref:`associative domains <primers-associative>` or sets). The Chapel standard
```
Comment (ShreyasKhandekar) [inline doc/rst/technotes/interfaces.rst:15]: Maybe mention the set module here when mentioning sets?

**Review by ShreyasKhandekar — APPROVED** [review]
Looks great! Just some grammar stuff :)

**Comment by DanilaFe** [comment]
Bleh, spellcheck doesn't catch invalid combinations of words. And apparently, I don't either when I read. Who knows how much of the stuff I've read is gibberish?

File: doc/rst/technotes/interfaces.rst (hunk 6 above)
Comment (DanilaFe) [reply doc/rst/technotes/interfaces.rst:15]: I think it gets a little bit too noisy in the introductory sentence (especially for a parenthesizes segment). There's another link to the `Set` module further down, so I don't feel too bothered by this.

**Pull request merged**